File: src/main/java/net/zero918nobita/Xemime/ast/Node.java
package net.zero918nobita.Xemime.ast;

import net.zero918nobita.Xemime.NodeType;
import net.zero918nobita.Xemime.Recognizable;
import net.zero918nobita.Xemime.entity.Bool;

import java.util.ArrayList;
import java.util.LinkedHashMap;

public class Node implements Recognizable {
    protected int location;  // 行番号

    public Node(int location) {
        this.location = location;
    }

    @Override
    public NodeType recognize() {
        return NodeType.NODE;
    }

    @Override
    public boolean is(NodeType comparison) {
        return comparison == recognize();
    }

    public int getLocation() {
        return this.location;
    }

    public Node run() throws Exception {
        return this;
    }

    public Node add(int location, Node rhs) throws FatalException {
        // `+` 演算子は使用できません
        throw new FatalException(location,  98);
    }

    public Node add(Node rhs) throws FatalException {
        return add(0, rhs);
    }

    public Node sub(int location, Node rhs) throws FatalException {
        // `-` 演算子は使用できません
        throw new FatalException(location, 99);
    }

    public Node sub(Node rhs) throws FatalException {
        return sub(0, rhs);
    }

    public Node multiply(int location, Node rhs) throws FatalException {
        // `*` 演算子は使用できません
        throw new FatalException(location,  100);
    }

    public Node multiply(Node rhs) throws FatalException {
        return multiply(0, rhs);
    }

    public Node divide(int location, Node rhs) throws FatalException {
        // `/` 演算子は使用できません
        throw new FatalException(location, 101);
    }

    public Node divide(Node rhs) throws FatalException {
        return divide(0, rhs);
    }

    public Node mod(int location, Node rhs) throws FatalException {
        // `%` 演算子は使用できません
        throw new FatalException(location, 140);
    }

    public Bool less(int location, Node rhs) throws FatalException {
        // `<` 演算子は使用できません
        throw new FatalException(location, 102);
    }

    public Bool less(Node rhs) throws FatalException {
        return less(0, rhs);
    }

    public Bool le(int location, Node rhs) throws FatalException {
        // `<=` 演算子は使用できません
        throw new FatalException(location, 103);
    }

    public Bool le(Node rhs) throws FatalException {
        return le(0, rhs);
    }

    public Bool greater(int location, Node rhs) throws FatalException {
        // `>` 演算子は使用できません。
        throw new FatalException(location, 104);
    }

    public Bool greater(Node rhs) throws FatalException {
        return greater(0, rhs);
    }

    public Bool ge(int location, Node rhs) throws FatalException {
        // `>=` 演算子は使用できません。
        throw new FatalException(location, 105);
    }

    public Bool ge(Node rhs) throws FatalException {
        return ge(0, rhs);
    }

    public Bool and(int location, Node rhs) throws FatalException {
        // `&&` 演算子は使用できません。
        throw new FatalException(location, 114);
    }

    public Bool and(Node rhs) throws FatalException {
        return and(0, rhs);
    }

    public Bool or(int location, Node rhs) throws FatalException {
        // `||` 演算子は使用できません。
        throw new FatalException(location, 115);
    }

    public Bool or(Node rhs) throws FatalException {
        return or(0, rhs);
    }

    public Bool xor(int location, Node rhs) throws FatalException {
        // `^` 演算子は使用できません。
        throw new FatalException(location, 116);
    }

    public Bool xor(Node rhs) throws FatalException {
        return xor(0, rhs);
    }

    public Node message(int location, Symbol symbol) throws Exception {
        throw new Exception(location + ": `" + toString() + "` にフィールドは設定できません");
    }

    public Node message(int location, Symbol symbol, LinkedHashMap<Symbol, Node> params) throws Exception {
        throw new Exception(location + ": `" + toString() + "` にメソッドは設定できません");
    }

    public Node message(int location, Symbol symbol, ArrayList<Node> params) throws Exception {
        throw new Exception(location + ": `" + toString() + "` にメソッドを設定できません");
    }
}

File: src/main/kotlin/Node.kt
package net.zero918nobita.xemime.ast

import net.zero918nobita.xemime.NodeType
import net.zero918nobita.xemime.Recognizable
import net.zero918nobita.xemime.entity.Bool
import java.util.ArrayList
import java.util.LinkedHashMap

open class Node(var location: Int) : Recognizable {
  @Throws(Exception::class)
  open fun run() = this

  @Throws(FatalException::class)
  open fun add(location: Int, rhs: Node): Node {
    throw FatalException(location, 98)
  }

  @Throws(FatalException::class)
  fun add(rhs: Node) = add(0, rhs)

  @Throws(FatalException::class)
  open fun sub(location: Int, rhs: Node): Node {
    throw FatalException(location, 99)
  }

  @Throws(FatalException::class)
  fun sub(rhs: Node) = add(0, rhs)

  @Throws(FatalException::class)
  open fun multiply(location: Int, rhs: Node): Node {
    throw FatalException(location, 100)
  }

  @Throws(FatalException::class)
  fun multiply(rhs: Node) = multiply(0, rhs)

  @Throws(FatalException::class)
  open fun divide(location: Int, rhs: Node): Node {
    throw FatalException(location, 101)
  }

  @Throws(FatalException::class)
  fun divide(rhs: Node) = divide(0, rhs)

  @Throws(FatalException::class)
  open fun mod(location: Int, rhs: Node): Node {
    throw FatalException(location, 140)
  }

  @Throws(FatalException::class)
  open fun less(location: Int, rhs: Node): Bool {
    throw FatalException(location, 102)
  }

  @Throws(FatalException::class)
  fun less(rhs: Node) = less(0, rhs)

  @Throws(FatalException::class)
  open fun le(location: Int, rhs: Node): Bool {
    throw FatalException(location, 103)
  }

  @Throws(FatalException::class)
  fun le(rhs: Node) = le(0, rhs)

  @Throws(FatalException::class)
  open fun greater(location: Int, rhs: Node): Bool {
    throw FatalException(location, 104)
  }

  @Throws(FatalException::class)
  fun greater(rhs: Node) = greater(0, rhs)

  @Throws(FatalException::class)
  open fun ge(location: Int, rhs: Node): Bool {
    throw FatalException(location, 105)
  }

  @Throws(FatalException::class)
  fun ge(rhs: Node) = ge(0, rhs)

  @Throws(FatalException::class)
  open fun and(location: Int, rhs: Node): Bool {
    throw FatalException(location, 114)
  }

  fun and(rhs: Node) = and(0, rhs)

  @Throws(FatalException::class)
  open fun or(location: Int, rhs: Node): Bool {
    throw FatalException(location, 115)
  }

  fun or(rhs: Node) = or(0, rhs)

  @Throws(FatalException::class)
  open fun xor(location: Int, rhs: Node): Bool {
    // `^` 演算子は使用できません。
    throw FatalException(location, 116)
  }

  @Throws(FatalException::class)
  fun xor(rhs: Node): Bool {
    return xor(0, rhs)
  }

  @Throws(Exception::class)
  open fun message(location: Int, symbol: Symbol): Node {
    throw Exception(location.toString() + ": `" + toString() + "` にフィールドは設定できません")
  }

  @Throws(Exception::class)
  open fun message(location: Int, symbol: Symbol, params: LinkedHashMap<Symbol, Node>): Node {
    throw Exception(location.toString() + ": `" + toString() + "` にメソッドは設定できません")
  }

  @Throws(Exception::class)
  open fun message(location: Int, symbol: Symbol, params: ArrayList<Node>): Node {
    throw Exception(location.toString() + ": `" + toString() + "` にメソッドを設定できません")
  }

  override fun recognize() = NodeType.NODE

  override fun `is`(comparison: NodeType) = comparison == recognize()
}
